### PR TITLE
Change pgAdmin Deployment to StatefulSet

### PR DESCRIPTION
**Pull request by andrewlecuyer**
Certain properties of a pgAdmin deployment make it better suited to be deployed via a StatefulSet rather than a Deployment (e.g. it includes its own embedded database, and we also want to ensure only a single replica is running at any given time).  This change therefore updates the intent generated by PGO for deploying pgAdmin to use a StatefulSet rather than a Deployment.  The StatefulSet uses the default `RollingUpdate` update strategy, and is also configured with a proper `ServiceName` as required for a stable network ID.

[sc-12915]

### CODE REVIEW

**Review by ValClarkson — APPROVED** [review]
lgtm

File: internal/controller/postgrescluster/pgadmin.go [hunk 1]
```diff
@@ -185,53 +185,58 @@ func (r *Reconciler) reconcilePGAdminDeployment(
...
 	// if the shutdown flag is set, set pgAdmin replicas to 0
 	if cluster.Spec.Shutdown != nil && *cluster.Spec.Shutdown {
-		deploy.Spec.Replicas = initialize.Int32(0)
+		sts.Spec.Replicas = initialize.Int32(0)
 	} else {
-		deploy.Spec.Replicas = cluster.Spec.UserInterface.PGAdmin.Replicas
+		sts.Spec.Replicas = cluster.Spec.UserInterface.PGAdmin.Replicas
 	}
 
 	// Don't clutter the namespace with extra ReplicaSets.
```
Comment (cbandy) [inline internal/controller/postgrescluster/pgadmin.go:193]: ⛏️ StatefulSet doesn't have ReplicaSets.

```suggestion
	// Don't clutter the namespace with extra ControllerRevisions.
```

File: internal/controller/postgrescluster/pgadmin.go [hunk 2]
```diff
@@ -185,53 +185,58 @@ func (r *Reconciler) reconcilePGAdminDeployment(
...
-		MaxUnavailable: intstr.ValueOrDefault(nil, intstr.FromInt(0)),
+	sts.Spec.RevisionHistoryLimit = initialize.Int32(0)
+
+	// Give the Pod a stable DNS record based on its name.
+	// - https://docs.k8s.io/concepts/workloads/controllers/statefulset/#stable-network-id
+	// - https://docs.k8s.io/concepts/services-networking/dns-pod-service/#pods
+	sts.Spec.ServiceName = naming.ClusterPodService(cluster).Name
+
+	// Set the StatefulSet update strategy to "RollingUpdate", and the Partition size for the
+	// update strategy to 0 (note that these are the defaults for a StatefulSet).
```
Comment (cbandy) [inline internal/controller/postgrescluster/pgadmin.go:202]: 🤔 I had to think for a moment about the partition "size" of zero. That made me think of an empty set and no-op. I think it's better to describe the effect of the values used.

```suggestion
	// Delete and recreate every pod of the StatefulSet when the pod template changes.
```

Should we also call out that known issue mentioned in the docs? Perhaps the link to the docs is enough?

> If you update the Pod template to a configuration that never becomes Running and Ready (for example, due to a bad binary or application-level configuration error), StatefulSet will stop the rollout and wait.

**Review by cbandy — APPROVED** [review]
Some nitpicks in comments.

File: internal/controller/postgrescluster/pgadmin.go (hunk 1 above)
Comment (andrewlecuyer) [reply internal/controller/postgrescluster/pgadmin.go]: Comment updated per suggestion.

File: internal/controller/postgrescluster/pgadmin.go (hunk 2 above)
Comment (andrewlecuyer) [reply internal/controller/postgrescluster/pgadmin.go]: Appended to the description to describe the effect, and also added a direct link to the "Forced rollback" section in the STS docs.

**Pull request merged**